Resync wallet action

benthecarman opened the pull request:
Creates a cli command:

```
$ wassabee dropwalletcache --wallet:walletName --network:walletNetwork
```
Closes https://github.com/zkSNACKs/WalletWasabi/issues/2005

yahiheb commented on WalletWasabi/Services/WalletService.cs:
```diff
@@ -253,6 +253,22 @@ private async void IndexDownloader_NewFilterAsync(object sender, FilterModel fil
+			if (File.Exists(TransactionsFilePath))
+			{
+				File.Delete(TransactionsFilePath);
+			}
```
`DataDir`,  `BlocksFolderPath`, `TransactionsFolderPath` and `TransactionsFilePath` should all be `camelCase`.

yahiheb replied:
Once #1920 gets merged it will be easier to detect this.

MaxHillebrand commented:
YES, concept ACK!

I like that it is right click, because it's not always needed and thus a button would be distracting.

As far as I understand your code changes, you delete the downloaded blocks, but keep the filters, then upon restart, it re-syncs the filters and re-downloads the blocks?

If my understanding is correct, then this is rather bandwidth intensive... Is there a more proper way of doing this without deleting and re-downloading the blocks, but just re-syncing the filters and downloading ONLY the missing blocks?

benthecarman commented:
> If my understanding is correct, then this is rather bandwidth intensive... Is there a more proper way of doing this without deleting and re-downloading the blocks, but just re-syncing the filters and downloading ONLY the missing blocks?

Possibly, I just recreated what was recommended to people in reddit threads. It might be possible to do it another way so we don't delete anything and rescan all of our filters.

benthecarman commented:
I get what you're saying @nopara73 but I still think it's better to add this while still trying to fix the wallet db. Doing this allows the UX to be better while we find the bugs as well as is an insurance if a bug comes up in the future.

MaxHillebrand commented:
I kinda agree with both @nopara73 and @benthecarman ...

I guess the question is, how long will it take to fix the architecture? If it takes longer than #twoweeks, then I think we need a temporary solution, and this PR is just that.

As soon as we fix the underlying issue, I think that this feature can be removed rather easily.

NicolasDorier commented:
I like having this, even if removed later on. I needed to rescan a lot.

nopara73 commented:
@lontivero what do you think?

lontivero commented:
IMO the ability to rescan a wallet could be something good to have but it is something ugly to have in the UI. I think this is one of those things that are candidate to have in the CLI. Something like: 

```
$ wasabi rescan --wallet:mywallet
```

benthecarman commented:
> IMO the ability to rescan a wallet could be something good to have but it is something ugly to have in the UI. I think this is one of those things that are candidate to have in the CLI.

I like this this idea, moved to a command as follows:

```
$ wassabee rescanwallet --wallet:walletName --network:walletNetwork
```

MaxHillebrand commented:
Testing on Qubes 4, Debian 9:

```
$ wassabee rescanwallet --wallet:HotSpending --network:testnet
wassabee: Unknown command: rescanwallet
wassabee: Use `wassabee help` for usage.
```

MaxHillebrand commented:
ah, [this](https://github.com/zkSNACKs/WalletWasabi/pull/2031#issuecomment-519781452) is a typo, the real command is `resyncwallet`!

```
$ wassabee resyncwallet --wallet:HotSpending --network:testnet
2019-08-09 15:27:10 INFO: Wallet loaded: HotSpending.
```

benthecarman commented:
> Concept ACK, not sure about the implementation. Did you try if it works? I think it shouldn't.

It works as in it deletes the files I expect it to.  The files I am deleting are based off what a reddit comment said. Should I be deleting the filters instead?

lontivero commented on WalletWasabi/Services/WalletService.cs:
```diff
@@ -253,6 +253,22 @@ private async void IndexDownloader_NewFilterAsync(object sender, FilterModel fil
+			string TransactionsFolderPath = Path.Combine(dataDir, "Transactions", network.ToString());
+			if (Directory.Exists(BlocksFolderPath))
+			{
+				Directory.Delete(BlocksFolderPath, true);
```
Do not delete the blocks, they could be hundreds and shared across multiple wallets. Also, take into account that in case the block file is corrupted for some reason it is re-downloaded.

MaxHillebrand replied:
Yes.
This is not just bandwidth intensive, it *might* also reduce network level privacy, when only these blocks are downloaded in the same time window. [The bandwidth issue is enough reason to not delete and re-download blocks]

lontivero commented:
```
$ dotnet run -- resyncwallet --wallet:TestNet --network:testnet                                                                          
2019-08-09 10:41:28 INFO: Wallet loaded: TestNet.
```

Maybe a user expects to see something in the screen. Basically the command name suggests that it will resync the wallet but what it really does is delete files. So, or well the command let the user know that next time he loads the wallet it will resync or, the command name should be changed to `dropwalletcachedata`.

lontivero commented on WalletWasabi.Gui/CommandLine/ResyncWalletCommand.cs:
```diff
@@ -0,0 +1,108 @@
+						error = true;
+					}
+					string dataDir = EnvironmentHelpers.GetDataDir(Path.Combine("WalletWasabi", "Client"));
+					WalletService.ResyncWallet(WalletName, dataDir, Network);
```
if `km` is null then it means it doesn't exist however we invoke the `ResyncWallet` anyway. Why?

nopara73 commented:
> It works as in it deletes the files I expect it to. The files I am deleting are based off what a reddit comment said. Should I be deleting the filters instead?

Yes, filters definitely. Is it enough? Maybe.

benthecarman commented:
- Renamed command to `dropwalletcache`

- Command now only deletes filters (tested this and it seems this is all that is necessary)

- Implemented proper error checking on the KeyManager

- Logged a success message so the output is not confusing to the user.

yahiheb commented on WalletWasabi.Gui/CommandLine/DropWalletCacheCommand.cs:
```diff
@@ -0,0 +1,112 @@
+				"eg: ./wassabee dropwalletcache --wallet:MyWalletName --network:main",
+				"",
+				{ "w|wallet=", "The name of the wallet file.", x =>  WalletName = x },
+				{ "n|network=", "The network for the given file (main, test, reg).", x => Network = GetNetwork(x)},
```
`GetNetwork(x)},` add space between `)` and `}`.

lontivero replied:
```suggestion
				{ "n|network=", "The network for the given file (main, test, reg). Default=mainnet", x => Network = GetNetwork(x)},
```

yahiheb commented on WalletWasabi.Gui/CommandLine/DropWalletCacheCommand.cs:
```diff
@@ -0,0 +1,112 @@
+				"",
+				{ "w|wallet=", "The name of the wallet file.", x =>  WalletName = x },
+				{ "n|network=", "The network for the given file (main, test, reg).", x => Network = GetNetwork(x)},
+				{ "h|help", "Show Help.", v => ShowHelp = true}
```
Same as above.

yahiheb commented on WalletWasabi/Services/WalletService.cs:
```diff
@@ -253,6 +253,16 @@ private async void IndexDownloader_NewFilterAsync(object sender, FilterModel fil
 
+		public static void DropWalletCache(string walletName, string dataDir, Network network)
+		{
+			string FiltersFolderPath = Path.Combine(dataDir, "IndexStore", network.ToString());
```
`FiltersFolderPath` should be camelCase.

benthecarman replied:
I do, I think this might have been a result of me copy-pasting.

lontivero commented:
Ok, this looks good. Just curious, how did you (@benthecarman) corrupt your wallet to test this?

benthecarman commented:
I didn't, I just put some Console.WriteLine()s in places that should be executing to make sure it's rescanning

lontivero commented:
This is what I did:

* Delete random compact filters (random row from the `IndexStore/MatureIndex.dat`)
* Delete random entries in the wallet blockstore section.
* Corrupted random entries in the wallet blockstore section (json remains valid)
* Switched random keys state from 2 (used) to 0 (clean) 

Then:
```
$ dotnet run
```
And open the corrupted wallet (TestNet.json in my case). I got:
```
2019-08-11 17:31:09 INFO: Wallet loaded: TestNet.
2019-08-11 17:31:15 WARNING IndexStore: Filters got corrupted. Reorging 100 immature filters in an attempt to fix them.
2019-08-11 17:31:17 INFO KeyManager: Corrected 100 heights.
2019-08-11 17:31:17 INFO: Start Chaumian CoinJoin service...
2019-08-11 17:31:17 INFO: Starting WalletService...
2019-08-11 17:31:17 INFO CcjClient: CcjClient is successfully initialized.
2019-08-11 17:31:22 INFO CcjClientState: Round (19357) added.
2019-08-11 17:31:22 INFO CcjClientState: Round (19358) added.
2019-08-11 17:31:24 INFO: WalletService started.
2019-08-11 17:31:54 ERROR WasabiSynchronizer: Inconsistent index state detected.
hashChain.TipHeight:1343118
hashChain.HashesLeft:230100
hashChain.TipHash:0000000000000161b203bb952ca9e3b80de6a540fd5f1ce1a7106073c1a30927
hashChain.HashCount:514544
hashChain.ServerTipHeight:1573218
firstFilter.BlockHash:000000000000021c266cc60a77f351d5232f76d512d5bfd1f8ce280bbfdfef63
firstFilter.BlockHeight:1573119
2019-08-11 17:31:54 CRITICAL IndexStore: Filters got corrupted and have no more immature filters.
2019-08-11 17:31:54 CRITICAL IndexStore: Deleting all filters and crashing the software...
```

Next I delete the cached data with:
```
$ dotnet run -- dropwalletcache --wallet:TestNet --network:testnet
2019-08-11 17:40:02 INFO: Wallet loaded: TestNet.
2019-08-11 17:40:02 INFO: Wallet cache successfully dropped!
$
```

Open Wasabi again and load the same wallet. It starts to synchronize again, downloading the filters and also some old blocks (this is unexpected because I already had all the blocks. Anyway, it is not related to this PR)

After synchronization is finished I have a huge mess in my wallet, **balance and coins are wrong**.
--------------

Finally I edited the wallet file and deleted the whole BlockStore json field and tried again. After that Wasabi displayed the correct balance and coins.

### Conclusion
Deleting the filters is not enough to fix a broken wallet state, we must also delete the cached state.

benthecarman commented:
It should now delete the clear the block state.

lontivero commented:
**Note:** I made a comment before approving this but somehow it was lost. 

I have tested this PR again with the latest version doing all what I did before (corrupting filters, keys and blockstate). After executing the new `dropwalletchache` command I launched Wasabi and opened the broken wallet. Wasabi began to synchronize and once the process was completed it still **displayed wrong balance and coins**. However, once I restarted Wasabi **everything looked well** again.

**Conclusion**
This new command effectively forces the wallet re-synchronization but the user needs to restart Wasabi in order to see the fixed state.

nopara73 commented on WalletWasabi/KeyManagement/KeyManager.cs:
```diff
@@ -817,7 +817,7 @@ public void AssertNetworkOrClearBlockState(Network expectedNetwork)
 					ToFileNoBlockchainStateLock();
 
-					if (lastNetwork != null)
+					if (lastNetwork != null && expectedNetwork != null)
```
`expectedNetwork` should not be null.

benthecarman replied:
I am using it to clear the current block state, by calling it with `expectedNetwork` as null. If this is not there then it will display a confusing log message.

nopara73 replied:
I know why you are doing that and I know it works. Yet it still smells. Let me elaborate on the larger picture here why I feel so strongly you should not do this: `AssertNetworkOrClearBlockState` was a hack by me in the first place. For some reason our state gets corrupted and I needed to hack this correcting function here. And now, you are hacking on top of my hack. It's a bad practice to use/rely on my hack in other parts of the code in the first place, but relying on my hack with another hack is really bad.

nopara73 commented on WalletWasabi/Services/WalletService.cs:
```diff
@@ -253,6 +253,16 @@ private async void IndexDownloader_NewFilterAsync(object sender, FilterModel fil
 
+		public static void DropWalletCache(string walletName, string dataDir, Network network)
+		{
+			string filtersFolderPath = Path.Combine(dataDir, "BitcoinStore", network.ToString(), "IndexStore");
```
This should not be written by hand. You should get this path somehow from the IndexStore.

benthecarman replied:
To do that I need to not make this function static and would also need to load the whole wallet service to run the command. I think this way would be cleaner.

lontivero replied:
Yep, it seems you're right. I don't know how we could get this from IndexStore.

nopara73 replied:
You need an initialized `Global`(`InitializeNoWalletAsync`). From that you can get the `BitcoinStore` and from that the path.

nopara73 replied:
Also take this function of the `WalletService`and add it to the `BitcoinStore`'s `IndexStore` itself. Its his responsibility. (Also don't make it static.) This way when we work into the `IndexStore`, let's say we change the files we'll know that we have to adjust this, too.

nopara73 replied:
What's be also nice if you'd create an `IClearMyState` interface which would have a `ClearState` function defined. This'd be applied to `IndexStore` and `KeyManager`. To sum up:

- `WalletService.DropWalletCache` -> `IndexStore.ClearState`
- `KeyManager.AssertNetworkOrClearBlockState` -> `KeyManager.AssertNetworkOrClearBlockState`+ `KeyManager.ClearState` where `AssertNetworkOrClearBlockState` may use the `ClearState` function.

benthecarman commented:
Ready for re-review @nopara73 @lontivero!

lontivero commented on WalletWasabi.Gui/CommandLine/DropWalletCacheCommand.cs:
```diff
@@ -0,0 +1,114 @@
+					}
+					else
+					{
+						await Daemon.Global.InitializeNoWalletAsync();
```
I am not sure about this, I would have prefer to let it as it was before because this does a lot of unnecessary things like starting Tor, the HWI process, the synchronizer, and the connection with peers. @nopara73, are you sure this is what you want?

nopara73 replied:
You are right, this smells, too. In this case, I don't know what's the clean solution is.

benthecarman replied:
We only really need to load the `BitcoinStore`, so we could copy the lines from the initalize function and do 

```c#
var BitcoinStore = new BitcoinStore();
await BitcoinStore.InitializeAsync(Path.Combine(Daemon.Global.DataDir, "BitcoinStore"), Network);
```

However, if we ever update the location or how this is done we would need to remember to change it in both places. We could make it it's own function though.

nopara73 replied:
Maybe elevate out the bitcoinstore initialization of the global, where nowallet initialization uses that and this uses that, too?

benthecarman commented:
Should be good now

nopara73 commented on WalletWasabi/KeyManagement/KeyManager.cs:
```diff
@@ -798,7 +798,7 @@ public void AssertNetworkOrClearBlockState(Network expectedNetwork)
 		public void ClearState()
 		{
-			BlockchainState.BestHeight = 0;
+			BlockchainState.BestHeight = Height.Unknown;
```
Why not 0? To be honest we don't really use `Height.Unknown` and I'm concerned if we start using it, it won't be handled properly.

benthecarman replied:
@lontivero commented to change it this

https://github.com/zkSNACKs/WalletWasabi/pull/2031#discussion_r316479159

nopara73 replied:
Ok, let's keep it like this.

nopara73 commented on WalletWasabi.Gui/Global.cs:
```diff
@@ -128,6 +128,13 @@ public async Task DesperateDequeueAllCoinsAsync()
+		{
+			BitcoinStore = new BitcoinStore();
+			var bstoreInitTask = BitcoinStore.InitializeAsync(Path.Combine(DataDir, "BitcoinStore"), network);
+			return bstoreInitTask;
```
Return with await here, because we always do that. However I was long wondering if we could just return `Task`s like this. Maybe it would worth looking it up what it causes and if it's better, then do it like this from now on.

benthecarman replied:
The reason I did it this way was because currently we initialize the task [here](https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi.Gui/Global.cs#L147), but do not call it with `await` until [here](https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi.Gui/Global.cs#L229)

nopara73 replied:
I don't get your point.

nopara73 replied:
Anyhow, as I said, it's not our coding convention, but it may be a good opportunity to discuss if this should be our new coding convention. It seems neither of us know or looked up what the code returning a task without awaiting does.

nopara73 commented:
Two more things those came to my mind. This is not the way to touch the index file.

1. You have to use the iomanager for that, that's the cross-process safe way to do it.
2. If this gets merged before PR https://github.com/zkSNACKs/WalletWasabi/pull/2188 then it has to be also added to that. This says something about the maintainability of this code.

benthecarman commented:
> If this gets merged before PR #2188 then it has to be also added to that. This foreshadows the unmaintainability of this code.

Will wait until it is merged then re-assess what to do with this.

jonathancross commented:
PS: #2188 was merged long ago + there is a conflict to be resolved.  Anything else holding this up?

benthecarman commented:
> PS: #2188 was merged long ago + there is a conflict to be resolved. Anything else holding this up?

I think I'll need to change a few things, I'll try to get it done in the next few days!

nopara73 commented:
https://github.com/zkSNACKs/WalletWasabi/issues/2359

benthecarman commented:
> #2359

Okay, this can wait until part 2 of it is done

benthecarman commented:
Is this still needed is should it be closed?

cc @nopara73

nopara73 commented:
Keep it open IMO

molnard commented:
Spent issue solved. We can close this IMO. @nopara73 what is your opinion?
